File: 0312-burst-balloons/0312-burst-balloons.cpp
class Solution {
public:
    int maxCoins(vector<int>& nums) {
        int n=nums.size();
        nums.push_back(1);
        nums.insert(nums.begin(), 1);
        vector<vector<int>>memo(n+1,vector<int>(n+1,-1));;
        int coins= calculateMaxCoins(1,n,nums,memo);
        return coins;
    }
    int calculateMaxCoins(int i, int j, vector<int>&newNums,  vector<vector<int>>&memo){
        if(i>j){
            return 0;
        }
        if(memo[i][j]!=-1){
            return memo[i][j];
        }
        int maxcoins=INT_MIN;
        for(int l=i;l<=j;l++){
            int c=newNums[i-1]*newNums[l]*newNums[j+1]+calculateMaxCoins(i,l-1,newNums,memo)+calculateMaxCoins(l+1,j,newNums,memo);
            maxcoins=max(maxcoins,c);
        }
        return memo[i][j]=maxcoins;
    }
};

File: 0312-burst-balloons/0312-burst-balloons.java
class Solution {
    public int maxCoins(int[] nums) {
        int n = nums.length;
        int[][] dp = new int[n+2][n+2];
        // add 1 to the left and right of the array
        int[] newNum = new int[n+2];
        newNum[0]=newNum[n+1]=1;
        for(int i=1;i<=n;i++){
            newNum[i]=nums[i-1]; // copying the elemenst of num into newNum
        }
        for(int length =1;length<=n;length++){
            for(int left = 1;left<=n-length+1;left++){
                int right = left+length-1;
                for(int i=left;i<=right;i++){
                    dp[left][right]= Math.max(dp[left][right],(newNum[left-1]*newNum[i]*newNum[right+1]) + dp[left][i-1]+dp[i+1][right]);
                }
            }
        }
        return dp[1][n];
        
        
    }
    
}